### PR TITLE
User/jf/add complete status file logic

### PR DESCRIPTION
**Pull request by feng-j678**
What is in this PR:
[X] rewrite logic to write status_payload to complete_status_file_path w/o truncation
[X] add unit test

### CODE REVIEW

File: src/core/src/bootstrap/Constants.py [hunk 1]
```diff
@@ -334,4 +336,11 @@ class UbuntuProClientSettings(EnumBackport):
         MINIMUM_PYTHON_VERSION_REQUIRED = (3, 5)  # using tuple as we can compare this with sys.version_info. The comparison will happen in the same order. Major version checked first. Followed by Minor version.
         MAX_OS_MAJOR_VERSION_SUPPORTED = 18
         MINIMUM_CLIENT_VERSION = "27.14.4"
+    class StatusTruncationConfig(EnumBackport):
+        INTERNAL_FILE_SIZE_LIMIT_IN_BYTES = 126 * 1024
+        AGENT_STATUS_FILE_SIZE_LIMIT_IN_BYTES = 128 * 1024
+        MIN_TRUNCATED_PACKAGE_COUNT = 5
+        TRUNCATION_ERROR_MESSAGE = "Results were truncated because too many patches were present, Check log for truncated packages"
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Constants.py:343]: nit: TRUNCATION_WARNING_MESSAGE = "Package lists were truncated to limit reporting data volume. In-VM logs contain complete lists."

File: src/core/src/service_interfaces/StatusHandler.py [hunk 2]
```diff
@@ -495,6 +496,10 @@ def __new_substatus_json_for_operation(operation_name, status="Transitioning", c
 
     # region - Status generation
     def __reset_status_file(self):
+        status_file_reset_content = json.dumps(self.__new_basic_status_json())
+        # Create complete status template
+        self.env_layer.file_system.write_with_retry(self.complete_status_file_path, '[{0}]'.format(status_file_reset_content), mode='w+')
+        # Create agent-facing status template
         self.env_layer.file_system.write_with_retry(self.status_file_path, '[{0}]'.format(json.dumps(self.__new_basic_status_json())), mode='w+')
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:503]: Why is line 503 not using the variable populated on line 499. 

The purpose of the variable is to avoid the json recreation twice.

File: src/core/src/service_interfaces/StatusHandler.py [hunk 3]
```diff
@@ -783,3 +780,24 @@ def __set_errors_json(self, error_count_by_operation, errors_by_operation):
...
+                else:
+                    self.composite_logger.log_error("Unable to read status file (retries exhausted). Error: {0}.".format(repr(error)))
+                    raise
+        return complete_status_data_raw
+
+    def __write_status_file(self, complete_status_payload):
+        truncated_status_payload = complete_status_payload
+        self.env_layer.file_system.write_with_retry_using_temp_file(self.status_file_path, '[{0}]'.format(json.dumps(truncated_status_payload)), mode='w+')
+
+    # endregion
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:803]: Why is this region marked Patch Truncation when none of the functions here are related to patch truncation

File: src/core/tests/Test_CoreMain.py [hunk 4]
```diff
@@ -65,7 +65,7 @@ def test_operation_fail_for_non_autopatching_request(self):
         self.__check_telemetry_events(runtime)
 
         # check status file
-        with runtime.env_layer.file_system.open(runtime.execution_config.status_file_path, 'r') as file_handle:
```
Comment (rane-rajasi) [inline src/core/tests/Test_CoreMain.py:68]: All the tests checking .status file should remain as is. 

The client/agent facing output is .status file and not .complete.status. Hence we need to check and assert things within .status file. 

Please revert the changes from status_file_path to complete_status_file_path in all tests

File: src/core/src/bootstrap/Constants.py [hunk 5]
```diff
@@ -335,3 +337,11 @@ class UbuntuProClientSettings(EnumBackport):
         MAX_OS_MAJOR_VERSION_SUPPORTED = 18
         MINIMUM_CLIENT_VERSION = "27.14.4"
 
+    class StatusTruncationConfig(EnumBackport):
+        INTERNAL_FILE_SIZE_LIMIT_IN_BYTES = 126 * 1024
+        AGENT_STATUS_FILE_SIZE_LIMIT_IN_BYTES = 128 * 1024
+        MIN_TRUNCATED_PACKAGE_COUNT = 5
+        TRUNCATION_WARNING_MESSAGE = "Package lists were truncated to limit reporting data volume. In-VM logs contain complete lists."
+
+
```
Comment (rane-rajasi) [inline src/core/src/bootstrap/Constants.py:346]: Why are these truncation constants being added in this PR? The PR is solely for adding a .complete.status file. There should not be any truncation related logic in here (that includes Constants). PRs should be completely independent on their own, adding truncation constants here adds dependency of the truncation PR on this one. And adds things that unrelated to the context of this PR.

File: src/core/src/service_interfaces/StatusHandler.py [hunk 6]
```diff
@@ -608,22 +601,22 @@ def load_status_file_components(self, initial_load=False):
...
+                    message = complete_status_file_data['status']['substatus'][i]['formattedMessage']['message']
                     self.__metadata_for_healthstore_summary_json = json.loads(message)
             if name == Constants.CONFIGURE_PATCHING_SUMMARY:     # if it exists, it must be to spec, or an exception will get thrown
                 if self.execution_config.exec_auto_assess_only:
-                    self.__configure_patching_substatus_json = status_file_data['status']['substatus'][i]
+                    self.__configure_patching_substatus_json = complete_status_file_data['status']['substatus'][i]
                 else:
-                    message = status_file_data['status']['substatus'][i]['formattedMessage']['message']
+                    message = complete_status_file_data['status']['substatus'][i]['formattedMessage']['message']
                     self.__configure_patching_summary_json = json.loads(message)
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:613]: L612 not tested

File: src/core/src/service_interfaces/StatusHandler.py [hunk 7]
```diff
@@ -660,22 +653,26 @@ def __write_status_file(self):
...
+            complete_status_payload['status']['substatus'].append(self.__metadata_for_healthstore_substatus_json)
         if self.__configure_patching_substatus_json is not None:
-            status_file_payload['status']['substatus'].append(self.__configure_patching_substatus_json)
-        if os.path.isdir(self.status_file_path):
+            complete_status_payload['status']['substatus'].append(self.__configure_patching_substatus_json)
+        if os.path.isdir(self.complete_status_file_path):
             self.composite_logger.log_error("Core state file path returned a directory. Attempting to reset.")
-            shutil.rmtree(self.status_file_path)
+            shutil.rmtree(self.complete_status_file_path)
+
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:670]: L669 not tested

File: src/core/src/service_interfaces/StatusHandler.py [hunk 8]
```diff
@@ -660,22 +653,26 @@ def __write_status_file(self):
...
+        # Write complete status file <seq.no>.complete.status
+        self.env_layer.file_system.write_with_retry_using_temp_file(self.complete_status_file_path, '[{0}]'.format(json.dumps(complete_status_payload)), mode='w+')
 
-        self.env_layer.file_system.write_with_retry_using_temp_file(self.status_file_path, '[{0}]'.format(json.dumps(status_file_payload)), mode='w+')
+        # Write agent status file
+        self.__write_status_file(complete_status_payload)
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:675]: Please use write_with_retry_using_temp_file() as it was used before. .status file is being accessed by both our extension and the agent. Writing to a temp file with retry was done in order to avoid file access and locking issues between the extension.

Also, an important note to consider in all PRs related to truncation, none of the existing functionality should change.

File: src/core/src/service_interfaces/StatusHandler.py [hunk 9]
```diff
@@ -783,3 +780,24 @@ def __set_errors_json(self, error_count_by_operation, errors_by_operation):
...
+            try:
+                with self.env_layer.file_system.open(file_path, 'r') as file_handle:
+                    complete_status_data_raw = json.load(file_handle)[0]    # structure is array of 1
+            except Exception as error:
+                if i < Constants.MAX_FILE_OPERATION_RETRY_COUNT - 1:
+                    time.sleep(i + 1)
+                else:
+                    self.composite_logger.log_error("Unable to read status file (retries exhausted). Error: {0}.".format(repr(error)))
+                    raise
+        return complete_status_data_raw
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:797]: There was a feedback about this in the original installation truncation PR: https://github.com/Azure/LinuxPatchExtension/pull/191#discussion_r1222142421

The comment applies here

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/tests/Test_CoreMain.py [hunk 10]
```diff
@@ -267,6 +267,7 @@ def test_assessment_operation_success(self):
         self.assertTrue(substatus_file_data[0]["name"] == Constants.PATCH_ASSESSMENT_SUMMARY)
         self.assertTrue(substatus_file_data[0]["status"].lower() == Constants.STATUS_SUCCESS.lower())
         self.assertTrue(substatus_file_data[1]["name"] == Constants.CONFIGURE_PATCHING_SUMMARY)
+        self.assertEqual(json.loads(substatus_file_data[1]['formattedMessage']['message'])['automaticOSPatchState'], 'Disabled')
```
Comment (rane-rajasi) [inline src/core/tests/Test_CoreMain.py:270]: Why are we checking automaticOSPatchState in this PR?

Comment (rane-rajasi) [reply src/core/tests/Test_CoreMain.py:270]: Applies to all other places this is being checked as well

File: src/core/tests/Test_StatusHandler.py [hunk 11]
```diff
@@ -396,6 +398,77 @@ def test_sort_packages_by_classification_and_state(self):
             self.assertEqual(installation_patches_sorted[12]["name"], "test-package-2")  # | Other              | Excluded    |
             self.assertEqual(installation_patches_sorted[13]["name"], "test-package-1")  # | Other              | NotSelected |
 
+    def test_if_complete_status_has_bad_json(self):
+        string_json = '[{"version": 1.0, "timestampUTC": "2023-05-13T07:38:07Z", "statusx": {"name": "Azure Patch Management", "operation": "Installation", "status": "success", "code": 0, "formattedMessage": {"lang": "en-US", "message": ""}, "substatusx": []}}]'
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:402]: Typo in the json: substatusx and statusx

Please verify all other fields too

File: src/core/tests/Test_StatusHandler.py [hunk 12]
```diff
@@ -396,6 +398,77 @@ def test_sort_packages_by_classification_and_state(self):
             self.assertEqual(installation_patches_sorted[12]["name"], "test-package-2")  # | Other              | Excluded    |
             self.assertEqual(installation_patches_sorted[13]["name"], "test-package-1")  # | Other              | NotSelected |
 
+    def test_if_complete_status_has_bad_json(self):
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:401]: This test is checking what happens when *.complete.status has an empty substatus. Empty substatus is not a bad json.
Please rename the function to test_complete_status_with_empty_substatus()

File: src/core/tests/Test_StatusHandler.py [hunk 13]
```diff
@@ -396,6 +398,77 @@ def test_sort_packages_by_classification_and_state(self):
...
+        self.assertTrue(len(json.dumps(substatus_file_data)) < self.test_internal_file_size_limit)
+        self.assertNotEqual(substatus_file_data["status"], Constants.STATUS_WARNING.lower())
+        self.assertEqual(len(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"]), patch_count_for_test )
+        status_file_patches = json.loads(substatus_file_data["formattedMessage"]["message"])["patches"]
+        self.assertNotEqual(status_file_patches[len(status_file_patches) - 1]['patchId'], "Truncated patch list record")
+        self.assertNotEqual(status_file_patches[len(status_file_patches) - 1]['name'], "Truncated patch list record")
+        self.assertEqual(json.loads(substatus_file_data["formattedMessage"]["message"])["errors"]["code"], 0)
+        self.assertEqual(len(json.loads(substatus_file_data["formattedMessage"]["message"])["errors"]["details"]), 0)
+        self.assertFalse("review this log file on the machine" in json.loads(substatus_file_data["formattedMessage"]["message"])["errors"]["message"])
+        # self.assertEqual(len(self.runtime.status_handler._StatusHandler__assessment_removed_packages), 0)
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:460]: Remove commented line

File: src/core/tests/Test_StatusHandler.py [hunk 14]
```diff
@@ -24,6 +25,7 @@
 
 class TestStatusHandler(unittest.TestCase):
     def setUp(self):
+        self.test_internal_file_size_limit = 126 * 1024
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:28]: Why is this size limit in the PR? This PR is creating *.complete.status file and copying it's contents to .status file. There is no logic on truncation in here. Please remove this and all it's related tests

File: src/core/tests/Test_StatusHandler.py [hunk 15]
```diff
@@ -396,6 +398,77 @@ def test_sort_packages_by_classification_and_state(self):
...
+            substatus_file_data = json.load(file_handle)[0]["status"]["substatus"]
+        self.assertIsNotNone(substatus_file_data)
+        self.assertEqual(len(substatus_file_data), 0)
+        self.runtime.env_layer.file_system.delete_files_from_dir(example_file1, "*.complete.status")
+
+    def test_if_complete_status_path_is_dir(self):
+        self.runtime.execution_config.status_file_path = self.runtime.execution_config.status_folder
+        self.assertRaises(Exception, self.runtime.status_handler.load_status_file_components(initial_load=True))
+
+    def test_write_truncated_status_file_under_size_limit(self):
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:425]: Please remove this test, this PR is not handling the truncation logic

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/tests/Test_StatusHandler.py [hunk 16]
```diff
@@ -397,6 +397,30 @@ def test_sort_packages_by_classification_and_state(self):
...
+        with self.runtime.env_layer.file_system.open(self.runtime.execution_config.status_file_path, 'r') as file_handle:
+            substatus_file_data = json.load(file_handle)[0]["status"]["substatus"]
+        self.assertIsNotNone(substatus_file_data)
+        self.assertEqual(len(substatus_file_data), 0)
+        self.runtime.env_layer.file_system.delete_files_from_dir(example_file1, "*.complete.status")
+
+    def test_if_complete_status_path_is_dir(self):
+        self.runtime.execution_config.status_file_path = self.runtime.execution_config.status_folder
+        self.assertRaises(Exception, self.runtime.status_handler.load_status_file_components(initial_load=True))
     def test_assessment_packages_map(self):
```
Comment (kjohn-msft) [inline src/core/tests/Test_StatusHandler.py:424]: missing new line between functions

File: src/core/tests/Test_StatusHandler.py [hunk 17]
```diff
@@ -397,6 +397,30 @@ def test_sort_packages_by_classification_and_state(self):
...
+        sample_json = '[{"version": 1.0, "timestampUTC": "2023-05-13T07:38:07Z", "statusx": {"name": "Azure Patch Management", "operation": "Installation", "status": "success", "code": 0, "formattedMessage": {"lang": "en-US", "message": ""}, "substatusx": []}}]'
+        file_path = self.runtime.execution_config.status_folder
+        example_file1 = os.path.join(file_path, '123.complete.status')
+        self.runtime.execution_config.complete_status_file_path = example_file1
+
+        with open(example_file1, 'w') as f:
+            f.write(sample_json)
+
+        status_handler = StatusHandler(self.runtime.env_layer, self.runtime.execution_config, self.runtime.composite_logger, self.runtime.telemetry_writer,
+            self.runtime.vm_cloud_type)
```
Comment (kjohn-msft) [inline src/core/tests/Test_StatusHandler.py:411]: nit:single line

File: src/core/tests/Test_StatusHandler.py [hunk 18]
```diff
@@ -397,6 +397,30 @@ def test_sort_packages_by_classification_and_state(self):
             self.assertEqual(installation_patches_sorted[12]["name"], "test-package-2")  # | Other              | Excluded    |
             self.assertEqual(installation_patches_sorted[13]["name"], "test-package-1")  # | Other              | NotSelected |
 
+    def test_if_complete_status_is_empty_json_with_bad_name(self):
```
Comment (kjohn-msft) [inline src/core/tests/Test_StatusHandler.py:400]: _test_if_status_file_resets_on_load_if_malformed_

better describes what is actually being validated to be true.

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Minor comments, okay on rest.

File: src/core/src/core_logic/ExecutionConfig.py [hunk 19]
```diff
@@ -74,10 +74,11 @@ def __init__(self, env_layer, composite_logger, execution_parameters):
 
         # Derived Settings
         self.log_file_path = os.path.join(self.log_folder, str(self.sequence_number) + ".core.log")
+        self.complete_status_file_path = os.path.join(self.status_folder, str(self.sequence_number) + ".complete" + ".status")
         self.status_file_path = os.path.join(self.status_folder, str(self.sequence_number) + ".status")
         self.include_assessment_with_configure_patching = (self.operation == Constants.CONFIGURE_PATCHING and self.assessment_mode == Constants.AssessmentModes.AUTOMATIC_BY_PLATFORM)
-        self.composite_logger.log_debug(" - Derived execution-config settings. [CoreLog={0}][StatusFile={1}][IncludeAssessmentWithConfigurePatching={2}]"
-                                        .format(str(self.log_file_path), str(self.status_file_path), self.include_assessment_with_configure_patching))
+        self.composite_logger.log_debug(" - Derived execution-config settings. [CoreLog={0}][CompleteStatusFile={1}][StatusFile={2}][IncludeAssessmentWithConfigurePatching={2}]"
```
Comment (rane-rajasi) [inline src/core/src/core_logic/ExecutionConfig.py:80]: [IncludeAssessmentWithConfigurePatching={**3**}

File: src/core/src/service_interfaces/StatusHandler.py (hunk 6 above)
Comment (rane-rajasi) [reply src/core/src/service_interfaces/StatusHandler.py:625]: Will not be covered in this PR

File: src/core/tests/Test_StatusHandler.py [hunk 20]
```diff
@@ -397,6 +397,30 @@ def test_sort_packages_by_classification_and_state(self):
...
+        # Mock complete status file is empty but with bad json name
+        sample_json = '[{"version": 1.0, "timestampUTC": "2023-05-13T07:38:07Z", "statusx": {"name": "Azure Patch Management", "operation": "Installation", "status": "success", "code": 0, "formattedMessage": {"lang": "en-US", "message": ""}, "substatusx": []}}]'
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:402]: Comment from the previous iteration still applies here: https://github.com/Azure/LinuxPatchExtension/pull/195#discussion_r1234566456

Please verify the exact names within json

File: src/core/tests/Test_StatusHandler.py [hunk 21]
```diff
@@ -397,6 +397,30 @@ def test_sort_packages_by_classification_and_state(self):
...
+        self.runtime.execution_config.complete_status_file_path = example_file1
+
+        with open(example_file1, 'w') as f:
+            f.write(sample_json)
+
+        status_handler = StatusHandler(self.runtime.env_layer, self.runtime.execution_config, self.runtime.composite_logger, self.runtime.telemetry_writer,
+            self.runtime.vm_cloud_type)
+
+        # Mock complete status file is empty but with bad json name and being called in the load_status_file_components, and it will recreate a good complete_status_file
+        status_handler.load_status_file_components(initial_load=True)
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:414]: q: Wouldn't StatusHandler init on line 410 also call load_status_file_components()? Why the repetitive call to this function again here?

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

**Review by kjohn-msft — APPROVED** [review]

File: src/core/src/service_interfaces/StatusHandler.py [hunk 22]
```diff
@@ -624,22 +613,36 @@ def load_status_file_components(self, initial_load=False):
...
+                    complete_status_file_data = json.load(file_handle)[0]    # structure is array of 1
+            except Exception as error:
+                if i < Constants.MAX_FILE_OPERATION_RETRY_COUNT - 1:
+                    time.sleep(i + 1)
+                else:
+                    self.composite_logger.log_error("Unable to read status file (retries exhausted). Error: {0}.".format(repr(error)))
+                    raise
+        return complete_status_file_data
+
+    def __write_complete_status_file(self):
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:645]: nit: the name __write_complete_status_file() suggests it only writes <>.complete.status file whereas this function writes both <>.complete.status and <>.status file

suggestion: rename this to __write_status_files()

File: src/core/tests/Test_StatusHandler.py (hunk 20 above)
Comment (rane-rajasi) [reply src/core/tests/Test_StatusHandler.py:402]: @feng-j678 this is still open

File: src/core/src/service_interfaces/StatusHandler.py (hunk 22 above)
Comment (rane-rajasi) [reply src/core/src/service_interfaces/StatusHandler.py:645]: @feng-j678, I'm approving this PR. It's okay if this PR goes in without this change, but please make sure to include this comment in your next PR

**Review by rane-rajasi — APPROVED** [review]
One minor comment left open. Leaving upto @feng-j678 whether or not to include the suggestion in this or next PR.

Branch shows out of sync with master, please resolve that before merging
